### ISSUE TITLE
cudaPackages: implement an allow(Newer|Older) mechanism

### ISSUE THREAD

**Issue by ConnorBaker** [NixOS/nixpkgs#266475]
> To clarify, it's not "break" in the sense that the derivation has broken = true -- I'm replacing builtins.throw which would "break" evaluation entirely.

> Right, we definitely do not want to `throw`. But `broken = true` is still an alternative to `optionalAttrs`, what do you think about those?

> _Originally posted by @SomeoneSerge in https://github.com/NixOS/nixpkgs/issues/266115#issuecomment-1800730353_

**Comment by samuela** [comment]
Is this currently blocking any packages/PRs? It looks like https://github.com/NixOS/nixpkgs/pull/266115 has already been merged

**Comment by ConnorBaker** [comment]
No, not blocking anything. Just had this thought since I’ve seen this pattern around as I’ve been rewriting `cuda-modules`. It could simplify things internally and perhaps make it easier for users to try out officially unsupported combinations without needing to do a number of deep overrides — I mean, assuming that’s something we’d want to let them do!

**Comment by SomeoneSerge** [comment]
I just had `torchvision` to even evaluate in https://github.com/NixOS/nixpkgs/pull/269639 because of https://github.com/NixOS/nixpkgs/blob/6fb31ec5f6b62e498aedb43f371d45075139ce06/pkgs/development/libraries/science/math/tensorrt/extension.nix#L27-L29, even though `torch`/`torchvision` doesn't actually use `tensorrt` by default.

I'm pretty sure `throw` is not the tool we need

**Comment by SomeoneSerge** [comment]
Some more reflection on the TRT situation: it's different from `cuda_compat`.

We only enable `cuda_compat` for `aarch64-linux` hosts, and we presume them to be jetsons. We do so because nvidia just chooses not to ship `cuda_compat` for the consumer gpus and x86_64 hosts, although in principle the same hack would totally work on any platform. Any recent cudatoolkit release clearly corresponds to a version of cuda_compat, so it "exists". Albeit only buildable on certain platforms.

With TRT, there are cudatoolkit releases for which TRT isn't a thing (yet). We couldn't even assign the broken/fake derivation any meaningful version. However, I still don't feel like omitting the attribute is a good thing, because it breaks eval so easily. If we leave out `cudaPackages.tensorrt`, we've got to remove `python3Packages.tensorrt` as well. I just don't feel like it's the kind of thing we want to maintain really?

**Comment by SomeoneSerge** [comment]
- [ ] `cuda-library-samples`

Going through https://github.com/NixOS/nixpkgs/pull/280386 I see that `cuda-library-samples` still contain `optionalAttrs` (e.g. they're confusingly missing in `nix-instantiate` in the linked PR) even after https://github.com/NixOS/nixpkgs/issues/273794
